### PR TITLE
Bug 1871059: Adding ability to wait for init containers to start and finish

### PR DESCRIPTION
**Pull request by shawn-hurley**
* Updating container status check to check the container state not readiness probe
* Adding warn condition on StageReport
* Handle stage pod not able to pull image.

### CODE REVIEW

File: pkg/controller/migmigration/stage.go [hunk 1]
```diff
@@ -482,24 +482,36 @@ func (t *Task) stagePodReport(client k8sclient.Client) (report PodStartReport, e
...
 		ready := false
 		for _, c := range pod.Status.ContainerStatuses {
-			if c.Ready {
+			if c.State.Running != nil {
+				// if a single container is running in the pod assume that the pod has started.
 				ready = true
 				break
 			}
+			if c.State.Waiting != nil {
+				progress = append(progress, c.State.Waiting.Message)
```
Comment (djwhatle) [inline pkg/controller/migmigration/stage.go:507]: @pranavgaikwad can you make sure this usage of progress makes sense

File: pkg/controller/migmigration/stage.go [hunk 2]
```diff
@@ -482,24 +482,36 @@ func (t *Task) stagePodReport(client k8sclient.Client) (report PodStartReport, e
 	}
 	report.started = true
 	for _, pod := range podList.Items {
+		initReady := true
+		for _, c := range pod.Status.InitContainerStatuses {
+			// If the init contianer is waiting, then nothing can happen.
+			if c.State.Waiting != nil {
```
Comment (alaypatel07) [inline pkg/controller/migmigration/stage.go:488]: hrm, I wonder if this is enough? what if the init container failed and is on its way to restart? 

Do we also need to check for `c.State.Terminated` to check for all non-running states and return messages?

Comment (shawn-hurley) [reply pkg/controller/migmigration/stage.go:489]: Updated

File: pkg/controller/migmigration/stage.go [hunk 3]
```diff
@@ -480,38 +481,112 @@ func (t *Task) stagePodReport(client k8sclient.Client) (report PodStartReport, e
...
+	err = sClient.List(context.TODO(), options, &podSList)
+
+	dPods := make(map[string]string)
+
+	for _, pod := range podDList.Items {
+		dPods[pod.Name] = pod.Name
+	}
+
+	for _, pod := range podSList.Items {
+		if _, exist := dPods[pod.Name]; !exist {
```
Comment (shawn-hurley) [inline pkg/controller/migmigration/stage.go:572]: @JaydipGabani This was the update that I made, see if it still works?

File: pkg/controller/migmigration/stage.go [hunk 4]
```diff
@@ -480,38 +481,112 @@ func (t *Task) stagePodReport(client k8sclient.Client) (report PodStartReport, e
...
-		for _, c := range pod.Status.ContainerStatuses {
-			if c.Ready {
-				ready = true
-				break
+		initReady := true
+		for _, c := range pod.Status.InitContainerStatuses {
+			// If the init contianer is waiting, then nothing can happen.
+			if c.State.Waiting != nil {
+				initReady = false
+				report.progress = append(report.progress, c.State.Waiting.Message)
```
Comment (pranavgaikwad) [inline pkg/controller/migmigration/stage.go:491]:
```suggestion
				report.progress = append(
				            report.progress,
				            fmt.Sprintf(
				                         "Pod %s/%s: Container %s %s",
				                         pod.Namespace,
				                         pod.Name,
				                         c.Name,
				                         c.State.Waiting.Message))
```

File: pkg/controller/migmigration/stage.go [hunk 5]
```diff
@@ -480,38 +481,112 @@ func (t *Task) stagePodReport(client k8sclient.Client) (report PodStartReport, e
...
+			report.started = true
+
+			return
+		}
+
+		// handle pod pending status
+		if pod.Status.Phase == corev1.PodPending {
+			for _, c := range pod.Status.ContainerStatuses {
+				if c.State.Waiting != nil {
+					report.progress = append(report.progress, c.State.Waiting.Message)
```
Comment (pranavgaikwad) [inline pkg/controller/migmigration/stage.go:526]:
```suggestion
				        report.progress = append(
				                    report.progress,
				                    fmt.Sprintf(
				                         "Pod %s/%s: Container %s %s",
				                         pod.Namespace,
				                         pod.Name,
				                         c.Name,
				                         c.State.Waiting.Message))
```

**Review by pranavgaikwad — CHANGES_REQUESTED** [review]

File: pkg/controller/migmigration/task.go [hunk 6]
```diff
@@ -474,6 +474,9 @@ func (t *Task) Run() error {
 		} else {
 			t.Requeue = PollReQ
 		}
+		if len(report.progress) > 0 {
```
Comment (pranavgaikwad) [inline pkg/controller/migmigration/task.go:477]: No need for a check, you can always just set an empty progress array and it will be omitted.

**Review by pranavgaikwad — APPROVED** [review]
Haven't tried this, but the code looks good to me. One redundant check on empty progress message can be removed.

File: pkg/controller/migmigration/stage.go [hunk 7]
```diff
@@ -480,38 +481,126 @@ func (t *Task) stagePodReport(client k8sclient.Client) (report PodStartReport, e
...
+			report.progress = append(report.progress, report.reasons...)
+			hasHealthyClaims(&pod)
+		}
+	}
+	return
+}
+
+// Match number of stage pods in source and destination cluster
+func (t *Task) allStagePodsMatch() (report []string, err error) {
+	dClient, err := t.getDestinationClient()
```
Comment (alaypatel07) [inline pkg/controller/migmigration/stage.go:557]: nit: we use `dstClient` and `srcClient` in most places if you want to follow the naming conventions

File: pkg/controller/migmigration/stage.go [hunk 8]
```diff
@@ -480,38 +481,126 @@ func (t *Task) stagePodReport(client k8sclient.Client) (report PodStartReport, e
...
+	if err != nil {
+		err = liberr.Wrap(err)
+		return
+	}
+
+	podDList := corev1.PodList{}
 
+	options := k8sclient.MatchingLabels(t.stagePodLabels())
+	err = dClient.List(context.TODO(), options, &podDList)
```
Comment (alaypatel07) [inline pkg/controller/migmigration/stage.go:566]: question: we seem to be listing stage pods across all namespaces, in order to reduce scope, should we just list in namespaces where the plan in running?

**Review by alaypatel07 — COMMENTED** [review]
Added a question probably and a nit, but apart from that LGTM. Huge +1 to the new `stagePodReport` func implementation

**Review by djwhatle — APPROVED** [review]
VISACK

File: pkg/controller/migmigration/stage.go (hunk 1 above)
Comment (djwhatle) [reply pkg/controller/migmigration/stage.go]: Pranav confirmed this was proper usage. 👍

File: pkg/controller/migmigration/stage.go (hunk 3 above)
Comment (JaydipGabani) [reply pkg/controller/migmigration/stage.go]: Works!

**Review by JaydipGabani — APPROVED** [review]

**Comment by shawn-hurley** [comment]
Rebasing and merging to commit attribution

**Pull request merged**